### DISCUSSION TITLE
computations do not agree WinPro, Ubuntu, MacOS

### DISCUSSION THREAD

**Discussion by fwessel** [magpylib/magpylib#392] [Q&A] [answered]
Any help is appreciated. 
I placed a circular coil symmetrically on the 6 faces of a cube, biased so that the fields cancel at the center of symmetry. 
The same code is run on three platforms: Win, Ubuntu, and Mac. The fields are measured at each coil center using getB for an observer. The Windows calculation looks right, with only a single component of field, by symmetry, but the other two do not (Linux Mac). Shown below are the observer positions and 3 field components. 

Python 3.9.7
WINDOWS 10 Pro, AMD Ryzen 9 16 core
C:\Users\fwessel\> python coiltest.py
(-500.0, 0, 0)[-2295.8844339 0.                 0.]
(500.0, 0, 0)[2295.8844339   0.                 0.]
(0, -500.0, 0)[0. 	-2.29588443e+03        0.]
(0, 500.0, 0)[0.		2.29588443e+03         0.]
(0, 0, -500.0)[0.          0.       -2.29588443e+03]
(0, 0, 500.0)[0.           0.        2.29588443e+03]

MacOS 11.6
(venv) pinch: fwessel$ python3 coiltest.py
(-500.0, 0, 0)[-2.29588443e+03  0.        -1.78885438e+03]
(500.0, 0, 0)[2.29588443e+03    0.         1.78885438e+03]
(0, -500.0, 0)[0.          -2.29588443e+03  1.78885438e+03]
(0, 500.0, 0)[0.            2.29588443e+03 -1.78885438e+03]
(0, 0, -500.0)[0.           0.             -2.29588443e+03]
(0, 0, 500.0)[0.            0.              2.29588443e+03]

**Comment by OrtnerMichael** [comment]
hi @fwessel ,

I think what you see is a numerical instability in the implementation. This happens when the observer is slightly off the coil axis (e.g. by 1e-15). Unfortunately this happens automatically when you rotate your coils around the observer point due to floating point errors - and my guess is they appear differently on different systems.

This problem is already in discussion in #374 and will be solved in the upcoming version 4.

To fix this problem now, you can move your observer position by just a little e.g. (1e-10, 1e-10, 1e-10) - then all systems should give the same results.

Does this solve your problem ?

**Comment by fwessel** [comment]
Hi mortner,
thanks i’ll check. 
fw

**Reply by fwessel** [reply]
I tried a number of offset permutations for one or more coils and for the observer, ranging from 1e-1 to 1e-16. The above results did not change noticeably.

**Comment by OrtnerMichael** [comment]
can you show your code ?

**Reply by fwessel** [reply]
```
from magpylib import Collection
from magpylib.current import Circular
import magpylib as mag3


#  physics variables
a = float(1E7)  # current
dia = 1000.0  # coil diameter
sep0 = dia/2.0  # coil placement
delta = float(1E-10)  #small dispacement
sep = sep0   #to correct for off-axis computational bug

#  current loop creation, superimpose loops and their fields
s1 = Circular(-a, dia).move([-sep + delta, 0, 0]).rotate_from_angax(90, [0, 1, 0])
s2 = Circular(a, dia).move([sep + delta, 0, 0]).rotate_from_angax(90, [0, 1, 0])

#  y displaced coils have opposite polarity for current from other coils
s3 = Circular(a, dia).move([0, -sep + delta, 0]).rotate_from_angax(90, [1, 0, 0])
s4 = Circular(-a, dia).move([0, sep + delta, 0]).rotate_from_angax(90, [1, 0, 0])

#  not necessary to rotate z displaced coils (xy plane)
s5 = Circular(-a, dia).move([0, 0, -sep + delta])
s6 = Circular(a, dia).move([0, 0, sep + delta])

c = Collection(s1, s2, s3, s4, s5, s6)


def bfield2(obsvr):  # grabs B-field from any location, location = obsvr
    # b = 0.
    sensor = mag3.Sensor(position=obsvr)
    b = mag3.getB(c, sensor)
    print(obsvr)
    print(b)
    # print(format(Decimal.from_float(b), '.2f'))

#  array of locations to be probed, centers of each coil and center of all coils

sep2 = sep + delta
loc_array = ((-sep + delta, 0, 0), (sep + delta, 0, 0), (0, -sep2 + delta, 0), (0, sep + delta, 0),
             (0, 0, -sep), (0, 0, sep2), (0, 0, 0))
for loc in loc_array:
    bfield2(loc)

# fig = plt.figure(figsize=(20,8))
# ax1 = fig.add_subplot(projection='3d')
# c.display(markers=[obsvr],axis=ax1) '# 3-d system'
# ax1.set_xlabel('x')
# ax1.set_ylabel('y')
# ax1.set_zlabel('z')
# plt.show()
```

**Comment by OrtnerMichael** [comment]
your implementation
```
loc_array = ((-sep + delta, 0, 0), (sep + delta, 0, 0), (0, -sep2 + delta, 0), (0, sep + delta, 0),
             (0, 0, -sep), (0, 0, sep2), (0, 0, 0))
```

doesn't move the observer points off-axis. You are just displacing along the axis. Please try something like

```
loc_array = ((-sep + delta, delta, delta), ...)
```

**Comment by fwessel** [comment]
OK:  that seems to works, but it seems to be a solution that is less than ideal. 

Does this “behavior”  manifest reliably for on-axis values with an “observer/sensor",
or does it also appear for general calculations of the field and for other specific conditions
that one needs to be aware of?  

I ask because in some of my particle trajectory calculations the volumetric-field magnitude 
seems to vary discontinuously.

BTW, any estimates when will the next release be available? 

Thanks.

**Comment by OrtnerMichael** [comment]
Glad that this one is understood - if not solved :).

This behavior manifests at numerical instabilities of the analytical implementation. They would typically appear in symmetry points and surface extensions. For the Current loop this means one has to be careful around positions at r=0, at r=r_loop and at z=0.

I understand that this is very problematic when considering volume integrals.

We already know how to solve these problems. V4.0.0 is still a bit off, but the stabilization of the current loop field is the next task on my agenda. So I can probably provide you with an installable pre-release where these issues are resolved for the current loop field within 1-2 weeks.

If this would be of help I will keep you updated on the progress.

**Comment by fwessel** [comment]
Understandable..I know the solns. at 
the points you mentioned are challenging. 
Your comments provide insight into 
the behavior I’ve been seeing. 

I’m looking forward to your beta. 
Thanks!
fw

**Comment by OrtnerMichael** [comment]
Hi, @fwessel 

just a status update. I managed to stabilize the field - it took a bit longer than anticipated. I will integrate the new expression into the development branch next week. I'll write you when its there - i hope its not oo late for you

sincerelym
michael

**Comment by fwessel** [comment]
Hi Michael, 
Thanks for the update…in the meantime we diverted to another project. 
We’re looking forward to testing the new module. 
Fw

**Comment by OrtnerMichael** [comment, marked as answer]
hi @fwessel,

so its working now on the branch [stability-and-v4-restructuing](https://github.com/magpylib/magpylib/tree/stability-and-v4-restructuing). I will merge this branch into [development](https://github.com/magpylib/magpylib/tree/development) once all tests are through.

This branch is based on the yet unpublished V4 so `.current.Circular` becomes now `.current.Loop`.

Please let me know if this fixes your problem. You should have 12-14 sig. figs. everywhere now :).

Also check out the new plotly plotting backend with  `display(object, backend='plotly')`

sincerely michael

**Comment by fwessel** [comment]
I am having a problem with the v4 code. I have tried a Helmholtz coil to
test out the magnetic calculations in my code, and I am only getting 2
dimensions of magnetic force for every c.getB calculation.
I am running on the 4.0.0 branch that you sent to fwessel:
magpylib               4.0.0
I changed my import as follows:

```python
from magpylib.current import Loop as Circular

#I did it this way so  that I wouldn't have to change all of my Circular
#text in the code.
#Here is how I set up the coils:

s7 = Circular(current=a,
diameter=dia).move([-(d/2),0,0]).rotate_from_angax(90, [0, 1, 0])
s8 = Circular(current=a,
diameter=dia).move([(d/2),0,0]).rotate_from_angax(90, [0, 1, 0])
c = Collection(s7,s8)

dia = 1000
a = 1.0e7
d = 1000
```

Please email me back at ***@***.*** if this is either an error on my
part, or if it is a bug in the code itself.
Thank you again for allowing us to use the development branch for MagPylib.

**Comment by OrtnerMichael** [comment]
hi @fwessel 

1. There was a field computation bug in the new implementation - it is fixed now on the development branch.
2. I do not understand what you mean with "_I am only getting 2 dimensions of magnetic force for every c.getB calculation._" please explain this if the patch did not solve your problem
3. Your Helmholtz coils are too far apart - they should be at distance R from each other - yours are at distance 2R.
4. your email is auto-hidden ***@***.***

**Comment by fwessel** [comment]
Hello again.
Iḿ comparing the current stable branch and the development branch of
magpylib by calculating field lines.
There still seems to be an inconsistency between the magpylib calculations
of the b-field and the code that I have been using to test field lines.
When I run the field line code on the current stable version of magpylib, I
get magImage.png as a result, whereas when I run the same code with the
development branch of magpylib, I get magImageV4-2.png as a result.
When I divide the b-field y and z components (c.getB()[1], c.getB[2]) by
450, the field lines come back into the magImage.png shape.
Please let me know if this is a problem in the scale factor or if I should
look more closely at my code.
I can also give you the code I use for the field lines if it would help.
Thank you

**Reply by Aquios7** [reply]
magImage.png
![magImage](https://user-images.githubusercontent.com/22086344/143338687-68b2f986-695c-44e8-a04e-b73bdb76ca65.png)
magImageV4-2.png
![magImageV4-2](https://user-images.githubusercontent.com/22086344/143338697-c08e6183-667e-429c-86c4-8b19c61b9aee.png)

**Reply by OrtnerMichael** [reply]
hi @fwessel

the problem is on our end. It seems that my solution is not as stable as I thought it is - thanks for pointing out the problem.

I'll check it out and get back to you.

**Comment by OrtnerMichael** [comment]
hi @fwessel 

I found the bug. You should now get a field that looks like this:

![image](https://user-images.githubusercontent.com/47384012/143655539-96788fa0-040c-4017-b121-b21555ad3cb2.png)

The fix is already implemented on the [development branch](https://github.com/magpylib/magpylib/tree/development).

Difficult to understand how our tests missed this - please let us know if this fixed your problems

**Reply by fwessel** [reply]
do you have a recommendation on how to plot the field lines in the  3D view, on the left-hand side?

**Reply by OrtnerMichael** [reply]
Myself not. I know that @Alexboiboi displays 3D field lines / iso lines in his Plotly implementations. Also @peter-leitner has experience with 3D B-field line computation - any input you two ?

**Reply by Alexboiboi** [reply]
Hi @fwessel, @OrtnerMichael ,

I opened an issue for 3D field line view (#417). I'll post some updates there or here directly

**Comment by Alexboiboi** [comment]
Hi @fwessel , @OrtnerMichael 

I wrote some helper functions, in order to ease the plotting procedure with `plotly`. It is nowhere near to be generalizable and it is still tricky to choose the right combination of parameters to make a descent plot. You can "play around" with the following code:

Note that depending on the parameters it can push the browser rendering to its limits quite quickly. 

Hope it helps nevertheless ;)

Best, Alex

```python
import magpylib as magpy
import numpy as np
import pandas as pd
import plotly.graph_objects as go

assert magpy.__version__ == '4.0.0'

# define streamtube helper functions

def validate_startplanes(val):
    """validate startplanes for plotly streamtube representation"""
    allowed_startplanes = ('x', 'y', 'z', 'xmin', 'ymin', 'zmin', 'xmax', 'ymax', 'zmax')
    assert all(v in allowed_startplanes for v in val.split('+')) or val in ('auto','all'), \
        f'startplanes must be a selection of {allowed_startplanes} or set to `auto`, ' \
        f'following values are invalid: \n{[v for v in val if v not in allowed_startplanes]}'
    return val

def process_startplanes_string(startplanes, df):
    """
    pre-processes the startplanes string for plotly streamtube representation 
        - 'all', all startplanes are used
        - 'auto', the dominant field direction x,y,z is computed and
    """
    if startplanes=='auto':
        idxmax = df[['u','v','w']].abs().mean().idxmax()
        startsaxis = ('x','y','z')[('u','v','w').index(idxmax)]
        startplanes = (startsaxis+'min', startsaxis+'max')
    elif startplanes=='all':
        startplanes=('xmin', 'ymin', 'zmin', 'xmax', 'ymax', 'zmax')
    else:
        startplanes = startplanes.split('+')
    return startplanes

def make_Streamtube(sources, pixel, startplanes='auto', normalize=False, field='getB', **kwargs):
    '''
    Create plotly Streamtube
    
    Parameters
    ----------
    sources: iterable of magpylib sources
        Bfield values need to be retrieved for the corresponding sources
    
    pixel: numpy array of shape (n,3)
        defines the positions of the vector field
    
    startplanes: str, default='auto'
        can be one of :
        - {'x', 'y', 'z', 'xmin', 'ymin', 'zmin', 'xmax', 'ymax', 'zmax'} or a
        combination thereof using `+` separator (e.g. `"xmin+xmax"`)
        - 'all', all startplanes are used
        - 'auto', the dominant field direction x,y,z is computed and 
    
    normalize: bool, default=False
        normalizes the vetor field
    
    field: 'getB' or 'getH'
        defines if the B-Field or H-Field should be used to compute the vector field
    
    Returns
    -------
    dict
        dictionary of compatible key/value pairs for a plotly Streamtube plot
    '''
    startplanes = validate_startplanes(startplanes)
    trace_dict = dict(type='streamtube')
    fieldfunc= getattr(magpy, field)
    Bvec = fieldfunc(sources, pixel, sumup=True)
    data = np.concatenate([pixel,Bvec],axis=1)
    df = pd.DataFrame(data=data, columns=['x','y','z','u','v','w']).dropna()
    if len(df.index)!=0:
        startplanes = process_startplanes_string(startplanes, df)
        masks = []
        for sp in startplanes:
            sp_axis, sp_arg = sp[0], sp[1:]
            df_filt = df[sp_axis]
            if sp_arg in ('min','max'):
                val = getattr(df_filt,sp_arg)()
            else:
                val = np.median(df_filt.unique())
            masks.append(~(df_filt==val))
        masks = pd.concat(masks, axis=1).all(axis=1)
        starts = df[~masks][['x','y','z']] \
            .rename(columns={k:'starts_'+k for k in df})
        if normalize:
            Bmag = np.linalg.norm(df[['u','v','w']].values, axis=1)
            df[['u','v','w']] = df[['u','v','w']]/np.tile(Bmag, (3,1)).T
        trace_dict.update(**df.to_dict('list'), **starts.to_dict('list'),
                            sizeref=0.2, showscale=False)
        trace_dict.update(**kwargs)
    return trace_dict

# define parameters
diameter = 1000
current = 1.0e7
gap = 1000

# define sources
loop1 = magpy.current.Loop(current=current, diameter=diameter, style_label="Loop 1") \
    .move([-(gap/2),0,0]).rotate_from_angax(90, [0, 1, 0])
loop2 = magpy.current.Loop(current=current, diameter=diameter, style_label="Loop 2") \
    .move([(gap/2),0,0]).rotate_from_angax(90, [0, 1, 0])

# define pixel positions
pix_dim = np.array([1,1,1])*gap*2
pix_Nelem = [10, 10, 10 ]
slices = [slice(-d/2,d/2,N*1j) for d,N in zip(pix_dim,pix_Nelem)]
pixel = np.mgrid[slices].reshape(len(slices),-1).T

# display sources on a plotly canvas
fig = go.Figure()
magpy.show(loop1, loop2, canvas=fig, backend='plotly')

# create Streamtube from sources, pixel and startplanes
streamtube_trace = make_Streamtube(sources=[loop1, loop2], pixel=pixel, startplanes='xmin+xmax', sizeref=0.1)

# add trace to canvas
fig.add_trace(streamtube_trace)

# show fig
fig.update_layout(height=600)
fig.show()
```

![image](https://user-images.githubusercontent.com/29252289/143859406-88f7929e-f49c-45ec-8349-2ef744d74f5a.png)

**Comment by fwessel** [comment]
I have found a small visual bug in the magpylib development branch.
Making a square out of the `magpy.current.Line` tool seems to have some of
the arrows going in the wrong direction.
Here is an image and my Line code for reference:

```python
# square coil in -x position
l1 = magpy.current.Line(current=current, vertices=([-500,-500,-500],
[-500,500,-500]))
l2 = magpy.current.Line(current=current, vertices=([-500,500,-500],
[-500,500,500]))
l3 = magpy.current.Line(current=current, vertices=([-500,500,500],
[-500,-500,500]))
l4 = magpy.current.Line(current=current, vertices=([-500,-500,500],
[-500,-500,-500]))
# square coil in x position
l5 = magpy.current.Line(current=current, vertices=([500,-500,-500],
[500,500,-500]))
l6 = magpy.current.Line(current=current, vertices=([500,500,-500],
[500,500,500]))
l7 = magpy.current.Line(current=current, vertices=([500,500,500],
[500,-500,500]))
l8 = magpy.current.Line(current=current, vertices=([500,-500,500],
[500,-500,-500]))

c = Collection(l1, l2, l3, l4, l5, l6, l7, l8)
```

**Reply by Aquios7** [reply]
![magImage_bug](https://user-images.githubusercontent.com/22086344/145907027-1a7530b0-fd52-4b4e-bed7-42504b8c321c.png)

**Comment by Alexboiboi** [comment]
Hi @fwessel,

I can confirm the bug, and I opened a Issue here: #429

Thank you for pointing this out.

to reproduce:
```python
import magpylib as magpy
import numpy as np

current = 10
vert1=np.array([
    [-500,-500,-500],
    [-500,500,-500],
    [-500,500,500],
    [-500,-500,500],
    [-500,-500,-500],
])
vert2 = vert1 + np.array((1000,0,0))
l1 = magpy.current.Line(current=current,vertices=vert1)
l2 = magpy.current.Line(current=current,vertices=vert2)

c = magpy.Collection(l1, l2)
c.display(backend='matplotlib')
```

**Reply by Alexboiboi** [reply]
Hi @fwessel,

* The issue is fixed and merged into the `development` branch with: #430 

fyi @OrtnerMichael

**Comment by fwessel** [comment]
Thank you. 
You guys are great!  
Onward and upward. 


(iPhone)
